zcash_primitives/src/legacy/keys.rs

```diff
@@ -0,0 +1,253 @@
+            .map(|k| k.private_key)
+    }
+}
+
+/// A type representing a BIP-44 public key at the account path level
+/// `m/44'/<coin_type>'/<account>'`.
+///
+/// This provides the necessary derivation capability for the for
+/// the transparent component of a unified full viewing key.
+#[derive(Clone, Debug)]
+pub struct AccountPubKey(ExtendedPubKey);
+
+impl AccountPubKey {
+    /// Derives the BIP-44 public key at the external "change level" path
+    /// `m/44'/<coin_type>'/<account>'/0`.
+    pub fn derive_external_ivk(&self) -> Result<ExternalIvk, hdwallet::error::Error> {
+        self.0
+            .derive_public_key(KeyIndex::Normal(0))
+            .map(ExternalIvk)
+    }
+
+    /// Derives the BIP-44 public key at the internal "change level" path
+    /// `m/44'/<coin_type>'/<account>'/1`.
+    pub fn derive_internal_ivk(&self) -> Result<InternalIvk, hdwallet::error::Error> {
+        self.0
+            .derive_public_key(KeyIndex::Normal(1))
+            .map(InternalIvk)
+    }
+
+    /// Derives the internal ovk and external ovk corresponding to this
+    /// transparent fvk. As specified in [ZIP 316][transparent-ovk].
+    ///
+    /// [transparent-ovk]: https://zips.z.cash/zip-0316#deriving-internal-keys
+    pub fn ovks_for_shielding(&self) -> (InternalOvk, ExternalOvk) {
+        let i_ovk = prf_expand_vec(
+            &self.0.chain_code,
+            &[&[0xd0], &self.0.public_key.serialize()],
+        );
+        let i_ovk = i_ovk.as_bytes();
+        let ovk_internal = InternalOvk(i_ovk[..32].try_into().unwrap());
+        let ovk_external = ExternalOvk(i_ovk[32..].try_into().unwrap());
+
+        (ovk_internal, ovk_external)
+    }
+
+    /// Derives the internal ovk corresponding to this transparent fvk.
+    pub fn internal_ovk(&self) -> InternalOvk {
+        self.ovks_for_shielding().0
+    }
+
+    /// Derives the external ovk corresponding to this transparent fvk.
+    pub fn external_ovk(&self) -> ExternalOvk {
+        self.ovks_for_shielding().1
+    }
+
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = self.0.chain_code.clone();
+        buf.extend(self.0.public_key.serialize().to_vec());
+        buf
+    }
+
+    pub fn deserialize(data: &[u8; 65]) -> Result<Self, hdwallet::error::Error> {
+        let chain_code = data[..32].to_vec();
+        let public_key = PublicKey::from_slice(&data[32..])?;
+        Ok(AccountPubKey(ExtendedPubKey {
+            public_key,
+            chain_code,
+        }))
+    }
+}
+
+/// Derives the P2PKH transparent address corresponding to the given pubkey.
+#[deprecated(note = "This function will be removed from the public API in an upcoming refactor.")]
+pub fn pubkey_to_address(pubkey: &secp256k1::key::PublicKey) -> TransparentAddress {
+    let mut hash160 = ripemd160::Ripemd160::new();
+    hash160.update(Sha256::digest(&pubkey.serialize()));
+    TransparentAddress::PublicKey(*hash160.finalize().as_ref())
```

Comment:
This can be slightly cleaned up (and switching to the non-deprecated `ripemd` crate) with something like:
```suggestion
    TransparentAddress::PublicKey(*ripemd::Ripemd160::digest(Sha256::digest(&pubkey.serialize()).as_ref()).as_ref())
```